### PR TITLE
A very simple interface for the very naive optimization functionality.

### PR DESCRIPTION
**Pull request by Ouyancheng**
A simple interface wrapper for the very naive optimization functionality.

### CODE REVIEW

**Comment by Ouyancheng** [comment]
@barrettcw @ajreynol

File: src/api/cpp/cvc5.h [hunk 1]
```diff
@@ -4155,14 +4265,48 @@ class CVC5_EXPORT Solver
...
+
+  /**
+   * Adds an optimization objective
+   * @param target the target formula to optimize
+   * @param objType whether it's MAXIMIZE or MINIMIZE
+   * @param bvSigned if the target is of sort bitvector,
+   *  indicate whether it's a signed bitvector or not, defaults to false,
+   *  for other types, it's omitted
+   */
+  void addObjective(Term target, ObjectiveType objType, bool bvSigned = false);
```
Comment (ajreynol) [inline src/api/cpp/cvc5.h:4280]: I'd prefer not adding this method, since it makes it so that objectives are part of the SMT state.

Can we make it so that checkOpt takes a vector of <Term, ObjectiveType> pairs?

Also, I don't think we should have a special flag bvSigned; can we make this part of ObjectiveType?  E.g. have 4 values { Min, Max, SignedMin, SignedMax } ?

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: src/api/cpp/cvc5.h (hunk 1 above)
Comment (Ouyancheng) [reply src/api/cpp/cvc5.h:4280]: I understand making it stateful will bring lots of problems in this case -- changing the assertions and other states might affect the optimization states (e.g., asserting new constraints between two runs of Pareto optimization will might get incorrect results), but I think that's inherent though -- if we want Pareto optimization return different valid solutions upon different runs, we need to make something as the state (even if we pass the objectives as arguments, we still need to store the objectives that we have solved previously, so it's also a state). And of course, handling either case would be clumsy anyways (and it's not well-handled yet)... 
As for making optimization objectives as states, it's definitely a design choice -- it doesn't bring us more trouble than handling the Pareto optimization AFAIK (WHEW... luckily) and gives us a possibility to make optimization incremental (same problem as handling Pareto). 
My thought is to make it like assertFormula / checkSat / getValue and addSygusConstraint / checkSynth / getSynthSolution triplets, to make it addObjective / checkOpt / getOptValue (not-yet there...) (and in the future if there's optimal synthesis, might be optSynth or synthOpt / getOptSynthSolution). 
What do you think of it? 

As for the objective types, I totally agree, and will work on it later on.

Comment (Ouyancheng) [reply src/api/cpp/cvc5.h]: It seems like Pareto + assertFormula in between is the only problem. Now it should handle Pareto correctly.

**Comment by Ouyancheng** [comment]
@barrettcw @ajreynol

**Comment by Ouyancheng** [comment]
@nafur @4tXJ7f

File: src/api/cpp/cvc5.h [hunk 2]
```diff
@@ -4372,6 +4486,40 @@ class CVC5_EXPORT Solver
...
+   *  indicate whether it's a BOX optimization,
+   *  a LEXICOGRAPHIC optimization or a PARETO optimization
+   * @return a pair <Result, vector<OptimizationResult>>,
+   *  where Result indicates the overall optimization result,
+   *  for BOX, if any of the objective is UNSAT or UNKNOWN,
+   *    it's UNSAT/UNKNOWN,
+   *  for LEXICOGRAPHIC, if the assertions are UNSAT/UNKNOWN,
+   *    it's UNSAT/UNKNOWN,
+   *  for PARETO, if the assertions are UNSAT
+   *  or the possible solutions are exhausted, it's UNSAT,
```
Comment (4tXJ7f) [inline src/api/cpp/cvc5.h:4505]: > or the possible solutions are exhausted, it's UNSAT

This makes it sound like the user should be calling this method multiple times to get the next solution. I don't think that this is what we want to suggest.

File: src/api/cpp/cvc5.h [hunk 3]
```diff
@@ -4372,6 +4486,40 @@ class CVC5_EXPORT Solver
...
+   *  vector<OptimizationResult> stores the results of individual objectives
+   */
+  std::pair<Result, std::vector<OptimizationResult>> checkOpt(
+      const std::vector<std::pair<Term, ObjectiveType>>& objectives,
+      ObjectiveCombination objCombination = LEXICOGRAPHIC) const;
+
+  /**
+   * Retrive the next Pareto solution,
+   * valid only after checkOpt is called with PARETO and SAT is returned,
+   * otherwise it's undefined behaviour.
```
Comment (4tXJ7f) [inline src/api/cpp/cvc5.h:4516]: The use of our API should never result in undefined behavior. We either throw an exception or we perform the action.

File: src/omt/bitvector_optimizer.h [hunk 4]
```diff
@@ -38,11 +38,11 @@ class OMTOptimizerBitVector : public OMTOptimizer
...
    * rounding towards -infinity: -1.5 --> -2 and 1.5 --> 1
    * same as the rounding scheme for int32_t
-   **/
+   */
   BitVector computeAverage(const BitVector& a,
                            const BitVector& b,
                            bool isSigned);
-  /** Is the BitVector doing signed comparison? **/
+  /** Is the BitVector doing signed comparison? */
   bool d_isSigned;
```
Comment (4tXJ7f) [inline src/omt/bitvector_optimizer.h:46]: It would be better not to include changes to unrelated files in PRs.

File: src/omt/integer_optimizer.h [hunk 5]
```diff
@@ -38,7 +38,7 @@ class OMTOptimizerInteger : public OMTOptimizer
    * Handles the optimization query specified by objType
    * isMinimize = true will trigger minimization,
    * otherwise trigger maximization
-   **/
+   */
   smt::OptimizationResult optimize(SmtEngine* optChecker,
                                    TNode target,
                                    bool isMinimize);
```
Comment (4tXJ7f) [inline src/omt/integer_optimizer.h:44]: It would be better not to include changes to unrelated files in PRs.

File: src/omt/omt_optimizer.h [hunk 6]
```diff
@@ -115,7 +115,7 @@ class OMTOptimizer
    * @param target the target expression to optimize
    * @return smt::OptimizationResult the result of optimization, containing
    *   whether it's optimal and the optimized value.
-   **/
+   */
   virtual smt::OptimizationResult maximize(SmtEngine* optChecker,
                                            TNode target) = 0;
 };
```
Comment (4tXJ7f) [inline src/omt/omt_optimizer.h:121]: It would be better not to include changes to unrelated files in PRs.

File: src/smt/optimization_solver.h [hunk 7]
```diff
@@ -221,7 +219,7 @@ class OptimizationSolver
   /**
    * Constructor
    * @param parent the smt_solver that the user added their assertions to
-   **/
+   */
```
Comment (4tXJ7f) [inline src/smt/optimization_solver.h:222]: It would be better not to include changes to the comments of unrelated methods in this PR.

File: src/api/cpp/cvc5.cpp [hunk 8]
```diff
@@ -6937,8 +6997,13 @@ std::string Solver::getOption(const std::string& option) const
...
 // Taken from https://en.cppreference.com/w/cpp/utility/variant/visit
-template<class... Ts> struct overloaded : Ts... { using Ts::operator()...; };
-template<class... Ts> overloaded(Ts...) -> overloaded<Ts...>;
+template <class... Ts>
+struct overloaded : Ts...
+{
+  using Ts::operator()...;
+};
+template <class... Ts>
+overloaded(Ts...) -> overloaded<Ts...>;
```
Comment (4tXJ7f) [inline src/api/cpp/cvc5.cpp:7006]: Please do not change the formatting of this in this PR.

File: src/api/cpp/cvc5.cpp [hunk 9]
```diff
@@ -3300,7 +3359,7 @@ std::vector<Term> Term::getSequenceValue() const
   //////// all checks before this line
   std::vector<Term> res;
   const Sequence& seq = d_node->getConst<Sequence>();
-  for (const auto& node: seq.getVec())
+  for (const auto& node : seq.getVec())
```
Comment (4tXJ7f) [inline src/api/cpp/cvc5.cpp:3362]: Please do not change the formatting of this in this PR.

File: src/api/cpp/cvc5.cpp [hunk 10]
```diff
@@ -1907,10 +1969,7 @@ size_t Op::getNumIndicesHelper() const
   return size;
 }
 
-Term Op::operator[](size_t index) const
-{
-  return getIndexHelper(index);
-}
+Term Op::operator[](size_t index) const { return getIndexHelper(index); }
```
Comment (4tXJ7f) [inline src/api/cpp/cvc5.cpp:1972]: Please do not change the formatting of this in this PR.

File: src/api/cpp/cvc5.cpp [hunk 11]
```diff
@@ -619,7 +620,7 @@ const static std::unordered_map<cvc5::Kind, Kind, cvc5::kind::KindHashFunction>
         {cvc5::Kind::BAG_IS_SINGLETON, BAG_IS_SINGLETON},
         {cvc5::Kind::BAG_FROM_SET, BAG_FROM_SET},
         {cvc5::Kind::BAG_TO_SET, BAG_TO_SET},
-        {cvc5::Kind::BAG_MAP,BAG_MAP},
+        {cvc5::Kind::BAG_MAP, BAG_MAP},
```
Comment (4tXJ7f) [inline src/api/cpp/cvc5.cpp:623]: Please do not change the formatting of this in this PR.

**Review by 4tXJ7f — CHANGES_REQUESTED** [review]
Some quick initial comments.

File: src/smt/optimization_solver.h (hunk 7 above)
Comment (Ouyancheng) [reply src/smt/optimization_solver.h:222]: oh I really don't want to add them back...

File: src/api/cpp/cvc5.h [hunk 12]
```diff
@@ -2901,6 +2994,27 @@ class CVC5_EXPORT Solver
...
 
  public:
+  /**
+   * Multiple-objective combination for optimization
+   */
+  enum ObjectiveCombination
+  {
+    BOX,
+    LEXICOGRAPHIC,
+    PARETO,
```
[the remark this answers is not in the document]
Comment (Ouyancheng) [reply src/api/cpp/cvc5.h:3004]: it should be in checkOpt.

File: src/api/cpp/cvc5.h (hunk 2 above)
Comment (Ouyancheng) [reply src/api/cpp/cvc5.h:4505]: forgot to update...

File: src/api/cpp/cvc5.h (hunk 3 above)
Comment (Ouyancheng) [reply src/api/cpp/cvc5.h:4516]: it's hard to detect whether the situation is ok or not. because you can add more assertions between the checkOpt and nextSolution and it's hard to detect it.

File: src/api/cpp/cvc5.cpp (hunk 11 above)
Comment (Ouyancheng) [reply src/api/cpp/cvc5.cpp:623]: I can revert these. But they should be formatted in another PR

File: src/api/cpp/cvc5.h [hunk 13]
```diff
@@ -272,6 +278,90 @@ std::ostream& operator<<(std::ostream& out, const Result& r) CVC5_EXPORT;
...
+   *   If the outcome is UNSAT or result is +-Inf,
+   *   it will be an empty term
+   **/
+  Term getValue() const;
+
+  /**
+   * Whether the result is +Inf/-Inf/Finite
+   * @return OptimizationResult::FINITE/POSITIVE_INF/NEGATIVE_INF
+   **/
+  IsInfinity isInfinity() const;
```
[the remark this answers is not in the document]
Comment (Ouyancheng) [reply src/api/cpp/cvc5.h:327]: it's like isSat in Result, instead of getting a bool it gets an enum...... (I don't really like this btw)

File: src/api/cpp/cvc5.h (hunk 3 above)
Comment (Ouyancheng) [reply src/api/cpp/cvc5.h:4516]: even worse, you can change the assertions between the two calls by performing a pop and re-asserting something

File: test/unit/theory/theory_opt_multigoal_white.cpp [hunk 14]
```diff
@@ -230,72 +230,72 @@ TEST_F(TestTheoryWhiteOptMultigoal, pareto)
...
-  results = optSolver.getValues();
-  ASSERT_EQ(results.size(), 0);
-}
+//   // resetting the assertions also resets the optimization objectives
+//   d_smtEngine->resetAssertions();
+//   r = optSolver.checkOpt(OptimizationSolver::LEXICOGRAPHIC);
+//   ASSERT_EQ(r.isSat(), Result::SAT);
+//   results = optSolver.getValues();
+//   ASSERT_EQ(results.size(), 0);
+// }
```
[the remark this answers is not in the document]
Comment (Ouyancheng) [reply test/unit/theory/theory_opt_multigoal_white.cpp]: uncommented

File: src/api/cpp/cvc5.cpp (hunk 8 above)
Comment (Ouyancheng) [reply src/api/cpp/cvc5.cpp]: stale

File: src/api/cpp/cvc5.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (Ouyancheng) [reply src/api/cpp/cvc5.h]: done (should be)

File: src/api/cpp/cvc5.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (Ouyancheng) [reply src/api/cpp/cvc5.h:3020]: oh should be inline instead

[the remark this answers is not in the document]
Comment (Ouyancheng) [reply src/api/cpp/cvc5.h:3044]: this is already updated, but GitHub doesn't show this idk why

**Review by ajreynol — CHANGES_REQUESTED** [review]
There are a few API methods that we did not agree on in this PR, please remove them from this PR.

Also, I would prefer if the internal changes to the optimization solver were done on a separate PR.

File: src/api/cpp/cvc5.h [hunk 15]
```diff
@@ -4480,6 +4685,18 @@ class CVC5_EXPORT Solver
...
+   * Adds an optimization objective
+   * @param target the target formula to optimize
+   * @param objType whether it's MAXIMIZE / MINIMIZE,
+   *   or BV_SIGNED_MAXIMIZE / BV_SIGNED_MINIMIZE
+   * Note: using MAXIMIZE/MINIMIZE on BitVector type implies
+   *   an **unsigned** optimization.
+   * Please use BV_SIGNED_MAXIMIZE / BV_SIGNED_MINIMIZE
+   *   for signed optimization.
+   */
+  void addObjective(Term target, ObjectiveType objType) const;
```
[the remark this answers is not in the document]
Comment (Ouyancheng) [reply src/api/cpp/cvc5.h:4698]: This is a private method and it's only for convenience.

File: src/api/cpp/cvc5.h [hunk 16]
```diff
@@ -4498,6 +4715,8 @@ class CVC5_EXPORT Solver
   std::unique_ptr<SmtEngine> d_smtEngine;
   /** The random number generator of this solver. */
   std::unique_ptr<Random> d_rng;
+  /** The optimization solver to handle optimization queries */
+  std::unique_ptr<smt::OptimizationSolver> d_optSolver;
```
[the remark this answers is not in the document]
Comment (Ouyancheng) [reply src/api/cpp/cvc5.h:4719]: oh, then let's not do API now before online optimization is implemented... 
I thought about online/inline optimizations before and I do think optimathsat being designed in that way is not for purely aesthetic reason, maybe some implementation difficulties force them so do. Let's find it out later

File: src/api/cpp/cvc5.h [hunk 17]
```diff
@@ -4372,6 +4547,36 @@ class CVC5_EXPORT Solver
...
+      ObjectiveCombination objCombination = LEXICOGRAPHIC) const;
+
+  /**
+   * Retrive the next Pareto solution,
+   * valid only after checkOpt is called with PARETO and SAT is returned.
+   * @return a pair <Result, vector<OptimizationResult>>,
+   *  where Result indicates the overall optimization result,
+   *  if the possible solutions are exhausted, it's UNSAT.
+   */
+  std::pair<Result, std::vector<OptimizationResult>> nextParetoSolution() const;
```
[the remark this answers is not in the document]
Comment (Ouyancheng) [reply src/api/cpp/cvc5.h:4578]: Sure, but see the above comment 👆

[the remark this answers is not in the document]
Comment (ajreynol) [reply src/api/cpp/cvc5.h:4578]: I don't follow. Can you explain?

[the remark this answers is not in the document]
Comment (Ouyancheng) [reply src/api/cpp/cvc5.h:4578]: the comment in the post above.

[the remark this answers is not in the document]
Comment (Ouyancheng) [reply src/api/cpp/cvc5.h:4578]: I mean this PR should be closed as I think we should figure out APIs only after the online optimization is implemented, instead of just arguing on the air.

[the remark this answers is not in the document]
Comment (ajreynol) [reply src/api/cpp/cvc5.h:4578]: Ok, makes sense.

[the remark this answers is not in the document]
Comment (ajreynol) [reply src/api/cpp/cvc5.h:4578]: We should discuss this in more detail.

In cvc5, we will need to extend the "SMT mode": https://github.com/cvc5/cvc5/blob/master/src/smt/smt_mode.h
This tracks the abstract state we are in on the diagram Figure 4.1 in page 52 of https://smtlib.cs.uiowa.edu/papers/smt-lib-reference-v2.6-r2021-05-12.pdf.
Notice that we have already informally extended this state diagram with two more modes "ABDUCT" and "INTERPOL".  The change of modes is managed at the SmtEngine level, e.g. see the comment here: https://github.com/cvc5/cvc5/blob/master/src/smt/smt_engine.cpp#L1719.

For optimization, we will likely need one new mode "SAT_OPTIMIZED".  An optimize-sat that returns "sat" goes to this mode, an "unsat" goes to the normal UNSAT mode.  The tricky part is that models *should be* available in "OPTIMIZED_SAT" mode; solutions for the last optimization problem are *only* available in "SAT_OPTIMIZED" mode.

[the remark this answers is not in the document]
Comment (ajreynol) [reply src/api/cpp/cvc5.h:4578]: As discussed in the meeting, we should not have the semantics of commands depend on options. This means that `check-sat` always means "check satisfibility without optimization" and `optimize-sat` means "check for satisfiability with optimization.  Moreover we provide the kind of optimization as an argument to `optimize-sat`.

[the remark this answers is not in the document]
Comment (Ouyancheng) [reply src/api/cpp/cvc5.h:4578]: but this does not incorporate well with (push) and (pop), and since the whole solver is in optimization mode we need to suppose this as well. as for the (check-sat), we can forbid this in optimization mode and force the user to use (optimize-sat). But supplying targets as arguments will definitely kill possibilities in (push) & (pop).

[the remark this answers is not in the document]
Comment (Ouyancheng) [reply src/api/cpp/cvc5.h:4578]: Oh I mean if we use (optimize-sat (target 1) (target 2)...), then it's hard to re-use the clauses next-time. 
of course we can implement an internal caching mechanism, but in this case every time the use does the pop we need to hook that and clear the corresponding cache.

[the remark this answers is not in the document]
Comment (ajreynol) [reply src/api/cpp/cvc5.h:4578]: Like with ordinary models, push and pop destroy the state of the current model, as it is no longer valid in the current context.  It also will destroy the state of the last optimization query; the objective has to be supplied again to another `optimize-sat` command.  We do not want to do anything fancier than this. 

Again, my recommendation is for us to focus on implementing the basic optimization methods without considering getting multiple solutions.

[the remark this answers is not in the document]
Comment (ajreynol) [reply src/api/cpp/cvc5.h:4578]: In particular, the important detail to notice is the edges marked "p" for push/pop in Figure 4.1 of https://smtlib.cs.uiowa.edu/papers/smt-lib-reference-v2.6-r2021-05-12.pdf.  This means that for:
```
(check-sat) ; returns "sat"
(pop)
...  the model is *not* available here.
```

[the remark this answers is not in the document]
Comment (Ouyancheng) [reply src/api/cpp/cvc5.h:4578]: sure, but the fact is that we have multiple objectives working, but the single query 😂 (we only have one offline BV binary search optimizer, but this is not enough, we should implement an online optimization for LIRA). 

for multiple objectives, it would be good to incorporate with push and pop, but if we don't want to it's ok (it's beneficial to lexico and box optimizations as we can use some of the learnt clauses, but it's hard to incorporate Pareto with push/pop).

[the remark this answers is not in the document]
Comment (Ouyancheng) [reply src/api/cpp/cvc5.h:4578]: > In particular, the important detail to notice is the edges marked "p" for push/pop in Figure 4.1 of https://smtlib.cs.uiowa.edu/papers/smt-lib-reference-v2.6-r2021-05-12.pdf. This means that for:
> 
> ```
> (check-sat) ; returns "sat"
> (pop)
> ...  the model is *not* available here.
> ```

Right right right, we need to check-sat again, but this time we used some of the learnt clauses in this level so it's faster. I think this should be incorporated into optimization as well so that we don't do everything again.
